doc: gsg: Edit instructions to set SDK env vars

greg-fer opened the pull request:
Edited the changes added in PR #46319.
The changes aim at including the variable info
in each OS installation procedure for major visibility.
Also added subsections on zephyr_sdk.rst page.

Signed-off-by: Grzegorz Ferenc <Grzegorz.Ferenc@nordicsemi.no>

greg-fer committed:
doc: gsg: Edit instructions to set SDK env vars

Edited the changes added in commit 0a758c3.
The changes aim at including the variable info
in a note before the OS installation procedure for major visibility.
Also added subsections on zephyr_sdk.rst page.

Signed-off-by: Grzegorz Ferenc <Grzegorz.Ferenc@nordicsemi.no>

stephanosio commented on doc/develop/getting_started/index.rst:
```diff
@@ -533,6 +533,12 @@ that are used to emulate, flash and debug Zephyr applications.
             sudo cp ~/zephyr-sdk-0.14.2/sysroots/x86_64-pokysdk-linux/usr/share/openocd/contrib/60-openocd.rules /etc/udev/rules.d
             sudo udevadm control --reload
 
+      #. Make sure the :envvar:`ZEPHYR_TOOLCHAIN_VARIANT` and :envvar:`ZEPHYR_SDK_INSTALL_DIR` environment variables
+         are set correctly (to ``zephyr`` and your Zephyr SDK installation directory, respectively).
+         See the :ref:`toolchain_zephyr_sdk_bundle_variables` section for details.
```
See https://github.com/zephyrproject-rtos/zephyr/pull/46319#discussion_r890599871 and https://github.com/zephyrproject-rtos/zephyr/pull/46319#discussion_r890602240

stephanosio commented on doc/develop/getting_started/index.rst:
```diff
@@ -580,6 +586,10 @@ that are used to emulate, flash and debug Zephyr applications.
             You must rerun the setup script if you relocate the Zephyr SDK bundle directory after
             the initial setup.
 
+      #. Make sure the :envvar:`ZEPHYR_TOOLCHAIN_VARIANT` and :envvar:`ZEPHYR_SDK_INSTALL_DIR` environment variables
+         are set correctly (to ``zephyr`` and your Zephyr SDK installation directory, respectively).
+         See the :ref:`toolchain_zephyr_sdk_bundle_variables` section for details.
```
same deal

greg-fer replied:
Edited as per comment https://github.com/zephyrproject-rtos/zephyr/pull/46365#discussion_r892052828

stephanosio commented on doc/develop/getting_started/index.rst:
```diff
@@ -621,11 +631,9 @@ that are used to emulate, flash and debug Zephyr applications.
-   You might need to set the :envvar:`ZEPHYR_TOOLCHAIN_VARIANT` and
-   :envvar:`ZEPHYR_SDK_INSTALL_DIR` environment variables. See the
-   :ref:`toolchain_zephyr_sdk` section for details.
+      #. Make sure the :envvar:`ZEPHYR_TOOLCHAIN_VARIANT` and :envvar:`ZEPHYR_SDK_INSTALL_DIR` environment variables
+         are set correctly (to ``zephyr`` and your Zephyr SDK installation directory, respectively).
+         See the :ref:`toolchain_zephyr_sdk_bundle_variables` section for details.
```
ditto

stephanosio commented on doc/develop/toolchains/zephyr_sdk.rst:
```diff
@@ -20,32 +23,39 @@ The Zephyr SDK supports the following target architectures:
-
-This allows the Zephyr build system to choose the correct version of the SDK,
-while allowing multiple Zephyr SDKs to be grouped together at a specific path.
+Toolchain variant variable
+   By default, the Zephyr build system assumes that the toolchain will be obtained
+   from the Zephyr SDK. You can enforce this by setting the environment variable
+   :envvar:`ZEPHYR_TOOLCHAIN_VARIANT` to ``zephyr``.
+
+SDK installation directory variable
```
I do not see how this is an improvement over the existing documentation.

It would indeed be useful to have an "environment variable glossary" page under `doc/develop/toolchains` that lists all the toolchain-related environment variables (referenced by `:envvar:`) with a detailed description, but what is done here seems like a half-baked solution.

mbolivar-nordic commented on doc/develop/getting_started/index.rst:
```diff
@@ -533,6 +533,11 @@ that are used to emulate, flash and debug Zephyr applications.
+         Make sure the :envvar:`ZEPHYR_TOOLCHAIN_VARIANT` and :envvar:`ZEPHYR_SDK_INSTALL_DIR` environment variables
+         are set correctly (to ``zephyr`` and your Zephyr SDK installation directory, respectively).
+         See the :ref:`toolchain_zephyr_sdk_bundle_variables` section for details.
+
```
Nothing wrong with this line, but please refer to the commit (0a758c36c864cc138941848f42c8ea4aa7dc95b3) instead of the pull request number (PR#46319) in the commit log.

Git is a decentralized version control system. Commits are "forever" and "global" objects across forks. Pull request numbers are specific to individual forks. So the log will be clearer if you use the commit instead. This also allows people to figure out the changes you are talking about even when they are offline.

greg-fer replied:
This is now applied, thanks!

mbolivar-nordic replied:
@greg-fer the commit log says:

```
Edited the changes added in commit #0a758c3.
```

it should say:

```
Edited the changes added in commit 0a758c3.
```

Commit SHAs are not numbers :).

mbolivar-nordic commented on doc/develop/getting_started/index.rst:
```diff
@@ -533,6 +533,11 @@ that are used to emulate, flash and debug Zephyr applications.
             sudo udevadm control --reload
 
+      .. note::
+         Make sure the :envvar:`ZEPHYR_TOOLCHAIN_VARIANT` and :envvar:`ZEPHYR_SDK_INSTALL_DIR` environment variables
```
Wait, why? These environment variables are optional these days; the default toolchain variant is `zephyr` if not set, and the installation directory is auto-discovered from a set of standard paths if not set. The "Make sure" language here suggests they are mandatory.

glarsennordic replied:
I like this edit, because for some projects, the `ZEPHYR_TOOLCHAIN_VARIANT` value was set to something other than `zephyr`, and when installing the new Zephyr toolchain, the old value is left untouched. This means the new install actually doesn't use the new toolchain, but uses whatever toolchain was previously installed. 

I think it is useful to have this mentioned as part of the gsg, because this is (indirectly) where a lot of users will be going to update their toolchain, and they (like me, at first) may not know or remember to even check the `ZEPHYR_TOOLCHAIN_VARIANT`

I agree, however, that the "Make sure" phrasing is misleading.

Perhaps better phrasing would be:

```rst
.. note::
        Make sure the :envvar:`ZEPHYR_TOOLCHAIN_VARIANT` and :envvar:`ZEPHYR_SDK_INSTALL_DIR` environment variables
        If you are installing the Zephyr toolchain to replace an existing Zephyr-compatible toolchain, check
        whether the :envvar:`ZEPHYR_TOOLCHAIN_VARIANT` and :envvar:`ZEPHYR_SDK_INSTALL_DIR` environment variables
        are set.
	
        If they are not set, you may proceed without making any changes (the new Zephyr toolchain will be selected
        by default).

        If they have a value, this will override Zephyr's default toolchain selection, and may prevent the Zephyr
        toolchain from being selected. In this case, either delete them, or set :envvar:`ZEPHYR_TOOLCHAIN_VARIANT`
	to ``zephyr`` and set :envvar:`ZEPHYR_SDK_INSTALL_DIR` to your Zephyr SDK installation directory. This
	will cause the Zephyr toolchain to be selected.
```

Though it is unclear to me what the ``zephyr`` toolchain variant actually does (does it cause the default selection process to occur, or does it specifically choose the Zephyr toolchain? Is "the Zephyr toolchain" even correct terminology, here?

greg-fer replied:
@glarsennordic , applied an edited version of your proposal *before* the installation steps. This is because the proposal reads like something you should be aware of before you start installing the SDK. Placing it before also allows to avoid duplicating the content in each tab.

mbolivar-nordic replied:
@glarsennordic 

The text

```
        If you are installing the Zephyr toolchain to replace an existing Zephyr-compatible toolchain, check
```

seems out of place for a document called the "Getting started guide".

A perennial problem that we face with this document is that everybody wants to include the little bits and pieces that they learned over time when adjusting their environment to fit their use cases, but these end up being noise for people who are really just getting started for the first time. We need to keep fighting to keep this document as short as possible, because installing zephyr is already complex enough.

glarsennordic commented:
Looks great -- I'm pretty sure my request for changes was not official since I am not a reviewer, but ping me if I am still somehow blocking merge

glarsennordic commented:
@stephanosio It looks like my change requests may have addressed your requested changes as well

Do you find the latest version satisfactory?

tejlmand commented on doc/develop/getting_started/index.rst:
```diff
@@ -476,6 +476,21 @@ Zephyr applications.
 It also contains additional host tools, such as custom QEMU and OpenOCD builds
 that are used to emulate, flash and debug Zephyr applications.
 
+.. note::
+   Before installing, check whether the :envvar:`ZEPHYR_TOOLCHAIN_VARIANT` or
+   :envvar:`ZEPHYR_SDK_INSTALL_DIR` environment variables are already set.
+
+   * If the variables are not set, the latest compatible version of the Zephyr SDK will be selected by default.
+     Proceed to next step without making any changes.
+
+   * If :envvar:`ZEPHYR_TOOLCHAIN_VARIANT` is set, the corresponding toolchain will be selected at build time
+     instead of the Zephyr SDK. Either unset the :envvar:`ZEPHYR_TOOLCHAIN_VARIANT` environment variable or change its
+     value to ``zephyr`` to make sure the Zephyr SDK is selected.
+
+   * If the :envvar:`ZEPHYR_SDK_INSTALL_DIR` environment variable is set, it will override the lookup location
+     for the selected toolchain. If you install the Zephyr SDK to one of the locations recommended below,
+     you can unset this variable. Otherwise, set it to your chosen install location.
```
I don't think this belongs here.

New users starting out with Zephyr or installing on a clean machine doesn't need to carry out those steps. They are imho just noise in those cases.

We should try to keep the guide as simple as possible without unnecessary steps for newcomers.

Instead this belongs in an *Updating Zephyr SDK* section or a trouble shooting section as this may only happen to users updating an existing system (which is beyond getting started).

If you add a *Update Zephyr SDK* section to the [Zephyr SDK](https://docs.zephyrproject.org/latest/develop/toolchains/zephyr_sdk.html) page then you can simple change the note to:
```suggestion
.. note::
   If you're updating the Zephyr SDK from an older release, please read :ref:`update_sdk`
``` 

Another option is to change the note to pure text and move it downwards to a *Troubleshooting* subsection. That subsection can then include the [asking-for-help](https://docs.zephyrproject.org/latest/develop/getting_started/index.html#asking-for-help) subsection.

This will be more aligned with what we have for [devicetree](https://docs.zephyrproject.org/latest/build/dts/index.html) and [west](https://docs.zephyrproject.org/latest/develop/west/index.html) where there exists dedicated [Troubleshooting](https://docs.zephyrproject.org/latest/develop/west/troubleshooting.html) sections for users experiencing problems.

greg-fer replied:
I went for adding a Troubleshooting section.

tejlmand commented on doc/develop/toolchains/zephyr_sdk.rst:
```diff
@@ -20,32 +23,39 @@ The Zephyr SDK supports the following target architectures:
-This allows the Zephyr build system to choose the correct version of the SDK,
-while allowing multiple Zephyr SDKs to be grouped together at a specific path.
+Toolchain variant variable
+   By default, the Zephyr build system assumes that the toolchain will be obtained
```
the build system doesn't assume anything.
It defaults to look for the Zephyr SDK if no toolchain has been selected.

How about:
```suggestion
If no toolchain is selected then the build system will look for Zephyr SDK and use a toolcahin from there.
```

tejlmand commented on doc/develop/toolchains/zephyr_sdk.rst:
```diff
@@ -20,32 +23,39 @@ The Zephyr SDK supports the following target architectures:
+
+SDK installation directory variable
+   If you install the Zephyr SDK outside any of the default locations (listed in
+   the operating system specific instructions below), you must register the Zephyr
```
> you must register

is this a **must** ?
And since when ?

greg-fer replied:
@carlescufi , this is a question for you :)

carlescufi replied:
It´s a must if you do not use environment variables @tejlmand?

In order for the SDK to be found by the build system you need:
- To install it in a default location OR
- to register it with CMake OR
- to set the `ZEPHYR_SDK_INSTALL_DIR`

isn't this correct?

tejlmand replied:
As long as you don't state how `ZEPHYR_SDK_INSTALL_DIR` is set.
It is correct that it must be set if not installed in default search path or registered as a CMake package.

But it doesn't have to be set in environment. It may be set in the CMake cache using `-D`.
Another option is to use the [Zephyr Build Configuration CMake package](https://docs.zephyrproject.org/latest/build/zephyr_cmake_package.html#zephyr-build-configuration-cmake-package), which can be used on project level to ensure a given toolchain is used. For example this is what is done by NCS and the NCS Toolchain.
Using a build configuration CMake package allows a project to define a specific toolchain and remove the need for user interaction.

Note: I just discovered that @greg-fer only reformatted the text and actually didn't change the wording.

How about this:

> If you install the Zephyr SDK outside any of the default locations (listed in the operating system specific instructions below) and you want automatic discovery of the Zephyr SDK, then you must register the Zephyr SDK in the CMake package registry by running the setup script.
> If you decide not to register the Zephyr SDK in the CMake registry, then the :envvar:`ZEPHYR_SDK_INSTALL_DIR` can be used to point to the Zephyr SDK installation directory.

That way we don't state that you must use one of those two ways, but that you can use one of those.
And at this point I don't believe we need to mention all possible options, as that should be generally describe, see https://github.com/zephyrproject-rtos/zephyr/issues/40389

greg-fer replied:
Applied suggestion...

tejlmand reviewed (changes requested):
Thanks for this.

I acknowledge that we can improve troubleshooting of Zephyr SDK updates, but I believe the **note** is misplaced, and I propose moving it to other location.

tejlmand commented:
@glarsennordic if you're satisfied with the latest changes, please approve the PR.

greg-fer commented:
Addressed some of your comments, @tejlmand , although I was not sure where to best place the Troubleshooting section.
One comment is pending answer from @carlescufi , as I've just restructured what he wrote.

tejlmand commented on doc/develop/getting_started/index.rst:
```diff
@@ -708,10 +702,35 @@ Here are some next steps for exploring Zephyr:
+* If the variables are not set, the latest compatible version of the Zephyr SDK will be selected by default.
+  Proceed to next step without making any changes.
+
+* If :envvar:`ZEPHYR_TOOLCHAIN_VARIANT` is set, the corresponding toolchain will be selected at build time
+  instead of the Zephyr SDK. Either unset the :envvar:`ZEPHYR_TOOLCHAIN_VARIANT` environment variable or change its
+  value to ``zephyr`` to make sure the Zephyr SDK is selected.
```
> corresponding toolchain will be selected at build time instead of the Zephyr SDK.

Well, only if the value is different from `zephyr` :wink: 

How about:
```suggestion
If :envvar:`ZEPHYR_TOOLCHAIN_VARIANT` is set, the corresponding toolchain will be selected at build time. Zephyr SDK is identified by the value `zephyr`.
If :envvar:`ZEPHYR_TOOLCHAIN_VARIANT` environment variable is not ``zephyr``, then either unset it or change its value to ``zephyr`` to make sure the Zephyr SDK is selected.
```

Note: this still only refers to the environment variable. In pratice the CMakeCache takes precedence, but I think we can wait for #40389 before cleaning up this.
For now, i'm satisfied with just referring to the envvar to keep things simple for the users.

tejlmand commented on doc/develop/getting_started/index.rst:
```diff
@@ -708,10 +702,35 @@ Here are some next steps for exploring Zephyr:
+  instead of the Zephyr SDK. Either unset the :envvar:`ZEPHYR_TOOLCHAIN_VARIANT` environment variable or change its
+  value to ``zephyr`` to make sure the Zephyr SDK is selected.
+
+* If the :envvar:`ZEPHYR_SDK_INSTALL_DIR` environment variable is set, it will override the lookup location
+  for the selected toolchain. If you install the Zephyr SDK to one of the locations recommended below,
```
Not sure I understand the meaning of this sentence:
>  If the :envvar:`ZEPHYR_SDK_INSTALL_DIR` environment variable is set, it will override the lookup location for the selected toolchain. 

So if I set `ZEPHYR_TOOLCHAIN_VARIANT` to `gnuarmemb`, then how will setting `ZEPHYR_SDK_INSTALL_DIR` impact the lookup of the `gnuarmemb` toolchain ?

greg-fer replied:
Summoning @glarsennordic  for this technical question :)

glarsennordic replied:
My apologies, I'm not actually 100% sure whether `ZEPHYR_SDK_INSTALL_DIR` applies only when `ZEPHYR_TOOLCHAIN_VARIANT="zephyr"` or if it impacts the lookup location for all of the toolchains.

> the selected toolchain

was sloppy phrasing on my part. It may be accurate, or it may not be. Someone actually familiar with the build configuration variables needs to chime in, and I don't know who that is.

If it only affects the zephyr SDK specifically, than better phrasing would be:

```rst
* If the :envvar:`ZEPHYR_SDK_INSTALL_DIR` environment variable is set, it will override the lookup location
  for the Zephyr SDK. If you install the Zephyr SDK to one of the locations recommended below,
```

Do we have any kind of reference documentation specifically dedicated to these environment variables, their function, and their possible values? That would help quite a bit.

greg-fer replied:
Applied edits.

tejlmand replied:
I just discovered `toolchain` has been changed to `Zephyr SDK` making the sentence clearer in meaning 👍 .
No further comments.

tejlmand reviewed (changes requested):
Thanks for creating a trouble shooting section.

Did a closer review at the text itself for that section.

greg-fer commented:
Applied some of your edits, @tejlmand , @carlescufi in a separate commit. I'll squash it after we nail the wording.

glarsennordic commented:
Approving with regard to the section in the gsg itself, with the major caveat that the statement about what `ZEPHYR_SDK_INSTALL_DIR` does may well be inaccurate. Someone familiar with these environment variables should review the final version of this section.

greg-fer commented:
@glarsennordic , @tejlmand , @carlescufi , please have another look.

- Modified "the selected toolchain" sentence as per feedback from Torsten and Georges.
- Moved the section to "Beyong GSG" page, as writing about updating made more sense there. I kept however the troubleshooting section with the section. Granted, there is some duplication, but I didn't think removing the info from GSG entirely was OK.
- Added links to "Important environmental variables" section, as per Georges' feedback.

mbolivar-nordic commented on doc/develop/env_vars.rst:
```diff
@@ -187,6 +187,8 @@ the :ref:`toolchain <gs_toolchain>` used to build Zephyr applications.
   ``ZEPHYR_TOOLCHAIN_VARIANT=llvm``, use :envvar:`LLVM_TOOLCHAIN_PATH`. (Note
   the capitalization when forming the environment variable name.)
 
+You might need to update some of these variables when you :ref:`update Zephyr SDK toolchain <gs_toolchain_update>`.
```
```suggestion
You might need to update some of these variables when you :ref:`update the Zephyr SDK toolchain <gs_toolchain_update>`.
```

Similar issue throughout -- shouldn't we say "the" Zephyr SDK toolchain, with a definite article?

mbolivar-nordic commented on doc/develop/getting_started/index.rst:
```diff
@@ -708,10 +702,27 @@ Here are some next steps for exploring Zephyr:
+.. _toolchain_zephyr_sdk_update:
+
+Double check Zephyr SDK variables when updating
+===============================================
+
+When updating Zephyr SDK, check whether the :envvar:`ZEPHYR_TOOLCHAIN_VARIANT` or :envvar:`ZEPHYR_SDK_INSTALL_DIR` environment variables are already set.
+See :ref:`gs_toolchain_update` for more information.
+
+For more information about these environmental variables in Zephyr, see :ref:`env_vars_important`.
+
 .. _help:
 
 Asking for Help
-***************
+===============
```
Why is this needed?

greg-fer replied:
OK, reverting.

mbolivar-nordic reviewed (changes requested):
Generally I am fine with adding a troubleshooting section.

However, this now seems to be making "Asking for help" a subsection of "troubleshooting". The commit message provides no justification as to why. My first instinct is to disagree with this change, as the two are different. I might not have any trouble to shoot at all with the basic getting started guide, but I would still want to know where to ask for help as I move along learning more about zephyr and running into problems with more advanced areas.

greg-fer commented:
@mbolivar-nordic , thanks for the suggestions! Applied all of them, and I agree about the "Ask for Help" section.

tejlmand reviewed (dismissed):
lgtm.

greg-fer commented:
@mbolivar-nordic , amended the commit description for hash "number". Please check the PR again if you have time.

greg-fer commented:
@tejlmand , looks like your approval has been dismissed with my push that amended the commit description. Could you approve again?

@glarsennordic , your check and approval would be also appreciated.

carlescufi reviewed (dismissed):
This looks good to me now, thanks @greg-fer!

stephanosio commented on doc/develop/beyond-GSG.rst:
```diff
@@ -74,6 +74,27 @@ setting :ref:`environment variables <env_vars>` such as
+  or change its value to ``zephyr`` to make sure Zephyr SDK is selected.
+
+* If the :envvar:`ZEPHYR_SDK_INSTALL_DIR` environment variable is set, it will override the lookup location
+  for Zephyr SDK. If you install Zephyr SDK to one of the locations recommended below,
```
> one of the locations recommended below

there aren't any locations listed **below**.

stephanosio reviewed (changes requested):
In general, please revise the formatting such that there is no more than 100 characters per line.

tejlmand commented on doc/develop/getting_started/index.rst:
```diff
@@ -708,6 +702,24 @@ Here are some next steps for exploring Zephyr:
+or :envvar:`ZEPHYR_SDK_INSTALL_DIR` environment variables are already set.
+See :ref:`gs_toolchain_update` for more information.
+
+For more information about these environmental variables in Zephyr, see :ref:`env_vars_important`.
```
```suggestion
For more information about these environment variables in Zephyr, see :ref:`env_vars_important`.
```

tejlmand reviewed (approved):
lgtm

mbolivar-nordic reviewed (approved):
The language is a bit inconsistent between "the Zephyr SDK" (which I think is right) and just plain "Zephyr SDK", but that's minor. Everything else LGTM.
